subprojects/docs/src/samples/providers/fileAndDirectoryProperty/kotlin/build.gradle.kts

```diff
@@ -1,30 +1,49 @@
-open class FooExtension(objects: ObjectFactory, layout: ProjectLayout) {
-    // A directory
-    val someDirectory: DirectoryProperty = objects.directoryProperty()
-    //  A file
-    val someFile: RegularFileProperty = objects.fileProperty()
-    // A collection of files or directories
-    val someFiles: ConfigurableFileCollection = layout.configurableFiles()
+// A project extension
+open class SourceGenerationExtension(objects: ObjectFactory) {
+    // The directory to write the generated source files to
+    val sourceDir: DirectoryProperty = objects.directoryProperty()
+
+    // The configuration file to use for source generation
+    val configFile: RegularFileProperty = objects.fileProperty()
 }
 
-project.extensions.create("foo", FooExtension::class, project.objects, project.layout)
+// A task that generates a source file and writes the result to an output directory
+open class GenerateSource @javax.inject.Inject constructor(objects: ObjectFactory): DefaultTask() {
```

Comment:
This is a lot of stuff on one line.
It might be better to do something like:
```kotlin
open class GenerateSource
@javax.inject.Inject
constructor(objects: ObjectFactory): DefaultTask() {
   ...
```